### PR TITLE
Language options

### PR DESCRIPTION
**Pull request by Eddasol**

### CODE REVIEW

**Comment by Eddasol** [comment]
Currently no functionality for the user to change language. To test Norwegian set 
"
   language: 'no',
   textDictionary: dictionaryList.no,
"
on line 15 and 16 in LanguageContext.tsx

File: frontend/src/language/no.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    "Name": "Navn",
+    "Completion Time": "Fullført tid",
+    "Please select asset": "Vennligst velg anlegg",
+    "This asset has no missions - Please create mission": "Dette anlegget har ingen oppdrag - Vennligst lag oppdrag",
+    "Cancel": "Avbryt",
+    "Schedule mission": "Planlegg oppdrag",
+    "Schedule Missions": "Planlegg oppdrag",
+    "Select robot": "Velg robot",
+    "Select start time": "Velg start tid",
+    "Cannot schedule mission in the past": "Kan ikke planglegge oppdrag i fortid",
```
Comment (GodVenn) [inline frontend/src/language/no.json:20]:
```suggestion
    "Cannot schedule mission in the past": "Kan ikke planlegge oppdrag i fortid",
```

File: frontend/src/language/no.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    "Status": "Status",
+    "Name": "Navn",
+    "Completion Time": "Fullført tid",
+    "Please select asset": "Vennligst velg anlegg",
+    "This asset has no missions - Please create mission": "Dette anlegget har ingen oppdrag - Vennligst lag oppdrag",
+    "Cancel": "Avbryt",
+    "Schedule mission": "Planlegg oppdrag",
+    "Schedule Missions": "Planlegg oppdrag",
+    "Select robot": "Velg robot",
+    "Select start time": "Velg start tid",
```
Comment (GodVenn) [inline frontend/src/language/no.json:19]:
```suggestion
    "Select start time": "Velg starttidspunkt",
```

File: frontend/src/language/no.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    "Delete mission": "Slett oppdrag",
+    "Time": "Tid",
+    "Date": "Dato",
+    "Estimated duration": "Estimert varighet",
+    "h": "t",
+    "min": "min",
+    "Estimated duration: not available": "Estimert varighet: ikke tilgjengelig",
+    "Create mission": "Lag oppdrag",
+    "No robot connected": "Ingen roboter er tilkoblet",
+    "Robot status": "Robot status",
```
Comment (GodVenn) [inline frontend/src/language/no.json:32]:
```suggestion
    "Robot status": "Robotstatus",
```

File: frontend/src/language/no.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    "Ongoing": "Pågår",
+    "Successful": "Vellykket",
+    "Aborted": "Avbrutt(Aborted)",
+    "Failed": "Feilet",
+    "Paused": "Pauset",
+    "Cancelled": "Avbrutt",
+    "Available": "Tilgjengelig",
+    "Busy": "Opptatt",
+    "Offline": "Offline",
+    "No robot position available": "Ingen robot posisjon tilgjengelig",
```
Comment (GodVenn) [inline frontend/src/language/no.json:43]:
```suggestion
    "No robot position available": "Ingen robotposisjon tilgjengelig",
```

File: frontend/src/language/no.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    "Robot position": "Robot posisjon",
```
Comment (GodVenn) [inline frontend/src/language/no.json:44]:
```suggestion
    "Robot position": "Robotposisjon",
```

File: frontend/src/language/no.json [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    "Start time": "Start tid",
```
Comment (GodVenn) [inline frontend/src/language/no.json:45]:
```suggestion
    "Start time": "Starttidspunkt",
```

File: frontend/src/language/no.json [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    "Time used": "Tid brukt",
+    "minutes": "minutter",
+    "Estimated time remaining": "Estimert gjenværende tid",
+    "Robot": "Robot",
+    "Battery level": "Batteri nivå",
```
Comment (GodVenn) [inline frontend/src/language/no.json:50]:
```suggestion
    "Battery level": "Batterinivå",
```

File: frontend/src/components/Contexts/LanguageContext.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export const LanguageContext = createContext<ILanguageContext>(defaultLanguage)
+
+export const LanguageProvider: FC<Props> = ({ children }) => {
+    const prevLanguage = window.localStorage.getItem('rcml-lang')
+    const [language, setLanguage] = useState(prevLanguage || defaultLanguage.language)
+
+    const textDictionary =
+        language === 'en' || language === 'no' ? dictionaryList[language] : defaultLanguage.textDictionary
```
Comment (Eddasol) [inline frontend/src/components/Contexts/LanguageContext.tsx:27]: Is there a better way to check if language is a valid key without hard-coding the keys?

**Review by GodVenn — COMMENTED** [review]
Et par kommentarer

File: frontend/src/components/Contexts/LanguageContext.tsx (hunk 8 above)
Comment (GodVenn) [reply frontend/src/components/Contexts/LanguageContext.tsx:27]: Maybe add config array for valid keys?

File: frontend/src/components/Contexts/LanguageContext.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    const switchLanguage = (newLanguage: string) => {
+        setLanguage(newLanguage)
+        window.localStorage.setItem('rcml-lang', newLanguage)
```
Comment (GodVenn) [inline frontend/src/components/Contexts/LanguageContext.tsx:30]: Should we name this `flotilla-language` instead if `rcml-lang`

File: frontend/src/components/Contexts/LanguageContext.tsx (hunk 8 above)
Comment (Eddasol) [reply frontend/src/components/Contexts/LanguageContext.tsx:27]: If I try to do the check using a list, I get an error:
image

or
image

**Review by aestene — APPROVED** [review]
LGTM

File: frontend/src/components/Contexts/LanguageContext.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
+import { createContext, FC, useContext, useState } from 'react'
+import { dictionaryList } from '../../language'
+
+interface ILanguageContext {
+    language: string
+    textDictionary: { [text: string]: string }
+    switchLanguage: (newLanguage: string) => void
+}
+
+interface Props {
```
Comment (phhuseboe) [inline frontend/src/components/Contexts/LanguageContext.tsx:10]: Props for what? Maybe consider more selfexplanatory name, for instance LanguageProviderProps

File: frontend/src/language/en.json [hunk 11]
```diff
@@ -0,0 +1,65 @@
+{
+    "Ongoing missions": "Ongoing missions",
+    "Upcoming missions": "Upcoming missions",
+    "Past Missions": "Past Missions",
```
Comment (phhuseboe) [inline frontend/src/language/en.json:4]: We should decide on capital letters in all header words, or just in the first word. 
I vote for capital letters in all header words, giving:

"Ongoing Missions": "Ongoing Missions",
"Upcoming Missions": "Upcoming Missions",
"Past Missions": "Past Missions"

The left and right hand side in the JSON should also be identical

File: frontend/src/language/en.json [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    "No ongoing missions": "No ongoing missions",
+    "No pending missions in queue": "No pending missions in queue",
+    "Status": "Status",
+    "Name": "Name",
+    "Completion Time": "Completion Time",
+    "Please select asset": "Please select asset",
+    "This asset has no missions - Please create mission": "This asset has no missions - Please create mission",
+    "Cancel": "Cancel",
+    "Schedule mission": "Schedule mission",
+    "Schedule Missions": "Schedule Missions",
```
Comment (phhuseboe) [inline frontend/src/language/en.json:17]: Should stay consequent whether it is Schedule missions or Schedule mission.
I propose missions in both, as it is possible to schedule multiple in one go.

Comment (Eddasol) [reply frontend/src/language/en.json:17]: Changing it to "Select missions" as the following lines are "Select robot" and "Select start time"

File: frontend/src/components/Contexts/LanguageContext.tsx (hunk 10 above)
Comment (Eddasol) [reply frontend/src/components/Contexts/LanguageContext.tsx:10]: Suggest keeping for consistent code between asset- and languagecontext

**Pull request merged**